projects/observability/src/shared/components/explore-query-editor/order-by/explore-query-order-by-editor.component.ts

```diff
@@ -0,0 +1,239 @@
+            (selectedChange)="this.onSortByChange($event)"
+          >
+            <ht-select-option
+              *ngFor="let option of sortByOptions"
+              [value]="option.value"
+              [label]="option.label"
+            ></ht-select-option>
+          </ht-select>
+        </div>
+      </div>
+    </div>
+  `
+})
+export class ExploreQueryOrderByEditorComponent implements OnChanges {
+  @Input()
+  public orderByExpression?: GraphQlOrderBy;
+
+  @Input()
+  public context?: TraceType;
+
+  @Output()
+  public readonly orderByExpressionChange: EventEmitter<GraphQlOrderBy | undefined> = new EventEmitter();
+
+  private readonly contextSubject: Subject<TraceType | undefined> = new ReplaySubject(1);
+  public readonly selectedMetric$: Observable<AttributeMetadata | undefined>;
+  public readonly selectedAggregation$: Observable<MetricAggregationType | undefined>;
+  public readonly selectedSortBy$: Observable<GraphQlSortDirection | undefined>;
+  public readonly metricOptions$: Observable<SelectOption<AttributeMetadata>[]>;
+  public readonly aggregationOptions$: Observable<SelectOption<MetricAggregationType>[]>;
+  public readonly sortByOptions$: Observable<SelectOption<GraphQlSortDirection | undefined>[]>;
+  private readonly incomingOrderByExpressionSubject: Subject<GraphQlOrderBy | undefined> = new ReplaySubject(1);
+
+  public constructor(private readonly metadataService: MetadataService) {
+    this.selectedAggregation$ = combineLatest([this.contextSubject, this.incomingOrderByExpressionSubject]).pipe(
+      switchMap(() => this.getCurrentSelectedAggregation())
+    );
+
+    this.selectedMetric$ = combineLatest([this.contextSubject, this.incomingOrderByExpressionSubject]).pipe(
+      switchMap(() => this.getCurrentSelectedAttribute())
+    );
+
+    this.selectedSortBy$ = this.incomingOrderByExpressionSubject.pipe(switchMap(() => this.getCurrentSelectedSortBy()));
+
+    this.metricOptions$ = this.contextSubject.pipe(switchMap(context => this.getAttributeOptionsForContext(context)));
+
+    this.aggregationOptions$ = this.selectedMetric$.pipe(
+      map(selection => this.getAggregationOptionsForAttribute(selection))
+    );
+
+    this.sortByOptions$ = this.getSortByOptions();
+  }
+
+  public ngOnChanges(changeObject: TypedSimpleChanges<this>): void {
+    if (changeObject.context) {
+      this.contextSubject.next(this.context);
+    }
+
+    if (changeObject.orderByExpression) {
+      this.incomingOrderByExpressionSubject.next(this.orderByExpression);
+    }
+  }
+
+  public onMetricChange(newAttribute: AttributeMetadata): void {
+    this.buildSpecAndEmit(
+      combineLatest([of(newAttribute), this.selectedAggregation$, of(this.orderByExpression?.direction)])
+    );
+  }
+
+  public onAggregationChange(newAggregation: MetricAggregationType): void {
+    this.buildSpecAndEmit(
+      combineLatest([this.selectedMetric$, of(newAggregation), of(this.orderByExpression?.direction)])
```

Comment:
Why are we building this strange observable and passing it to a function? I can't imagine how this will work without a subscribe later. Code smell here, should try to refactor to pass in sync values. Will read on.

Reply:
It has a subscription later. So I will keep i.